File: services/validation.py
"""
Input validation and proposal state machine.
"""

import re
import time
from collections import defaultdict
from typing import Optional, Tuple

# ---------------------------------------------------------------------------
# Proposal state machine -- valid transitions
# ---------------------------------------------------------------------------
VALID_TRANSITIONS = {
    "draft": {"pending_review"},
    "pending_review": {"approved", "revision_requested", "rejected"},
    "revision_requested": {"pending_review", "rejected"},
    "approved": {"executing"},
    "executing": {"completed", "failed"},
    "completed": set(),
    "failed": {"pending_review"},
    "rejected": set(),
}

VALID_PROPOSAL_TYPES = {
    "campaign", "landing_page", "ad_copy", "budget", "strategy", "content", "general",
}

VALID_LEAD_STATUSES = {
    "new", "contacted", "qualified", "consultation_booked", "converted", "lost",
}


def validate_status_transition(current: str, new: str) -> Tuple[bool, str]:
    """Check if a proposal status transition is valid."""
    allowed = VALID_TRANSITIONS.get(current, set())
    if new in allowed:
        return True, ""
    return False, f"Invalid transition: {current} -> {new}. Allowed: {', '.join(allowed) or 'none'}"


def validate_lead_input(data: dict) -> Tuple[bool, list]:
    """Validate lead creation input. Returns (valid, errors)."""
    errors = []
    name = (data.get("name") or "").strip()
    phone = (data.get("phone") or "").strip()
    email = (data.get("email") or "").strip()

    if not name or len(name) < 2:
        errors.append("שם חייב להכיל לפחות 2 תווים")

    if not phone:
        errors.append("טלפון הוא שדה חובה")
    elif not re.match(r'^[\d\-\+\s\(\)]{7,20}$', phone):
        errors.append("מספר טלפון לא תקין")

    if email and not re.match(r'^[^@\s]+@[^@\s]+\.[^@\s]+$', email):
        errors.append("כתובת אימייל לא תקינה")

    return (len(errors) == 0, errors)


def validate_proposal_input(data: dict) -> Tuple[bool, list]:
    """Validate proposal creation input."""
    errors = []
    title = (data.get("title") or "").strip()
    ptype = (data.get("proposal_type") or "").strip()

    if not title:
        errors.append("Title is required")
    if ptype and ptype not in VALID_PROPOSAL_TYPES:
        errors.append(f"Invalid proposal_type: {ptype}. Allowed: {', '.join(VALID_PROPOSAL_TYPES)}")

    return (len(errors) == 0, errors)


# ---------------------------------------------------------------------------
# Simple in-memory rate limiter
# ---------------------------------------------------------------------------
class RateLimiter:
    """Token-bucket rate limiter per key (e.g., IP or endpoint)."""

    def __init__(self, max_requests: int = 10, window_seconds: int = 60):
        self.max_requests = max_requests
        self.window = window_seconds
        self._hits = defaultdict(list)

    def is_allowed(self, key: str = "global") -> bool:
        now = time.time()
        self._hits[key] = [t for t in self._hits[key] if t > now - self.window]
        if len(self._hits[key]) >= self.max_requests:
            return False
        self._hits[key].append(now)
        return True


# Singleton instances
chat_limiter = RateLimiter(max_requests=10, window_seconds=60)
api_limiter = RateLimiter(max_requests=30, window_seconds=60)

File: lib/validation.ts
import { z } from "zod";

export const VALID_LEAD_STATUSES = [
  "new",
  "contacted",
  "qualified",
  "consultation_booked",
  "converted",
  "lost",
] as const;

export const VALID_PROPOSAL_TYPES = [
  "campaign",
  "landing_page",
  "ad_copy",
  "budget",
  "strategy",
  "content",
  "general",
] as const;

export const VALID_PROPOSAL_STATUSES = [
  "draft",
  "pending_review",
  "approved",
  "revision_requested",
  "rejected",
  "executing",
  "completed",
  "failed",
] as const;

const VALID_TRANSITIONS: Record<string, Set<string>> = {
  draft: new Set(["pending_review"]),
  pending_review: new Set(["approved", "revision_requested", "rejected"]),
  revision_requested: new Set(["pending_review", "rejected"]),
  approved: new Set(["executing"]),
  executing: new Set(["completed", "failed"]),
  completed: new Set(),
  failed: new Set(["pending_review"]),
  rejected: new Set(),
};

export function validateStatusTransition(
  current: string,
  next: string
): { valid: boolean; error: string } {
  const allowed = VALID_TRANSITIONS[current] ?? new Set();
  if (allowed.has(next)) return { valid: true, error: "" };
  const allowedStr = Array.from(allowed).join(", ") || "none";
  return {
    valid: false,
    error: `Invalid transition: ${current} -> ${next}. Allowed: ${allowedStr}`,
  };
}

export const leadInputSchema = z.object({
  name: z.string().min(2, "שם חייב להכיל לפחות 2 תווים"),
  phone: z
    .string()
    .min(1, "טלפון הוא שדה חובה")
    .regex(/^[\d\-+\s()]{7,20}$/, "מספר טלפון לא תקין"),
  email: z
    .string()
    .email("כתובת אימייל לא תקינה")
    .optional()
    .or(z.literal("")),
  source: z.string().default("manual"),
  campaign: z.string().optional().default(""),
  location: z.string().optional(),
  budget: z.string().optional(),
  projectType: z.string().optional(),
  notes: z.string().optional(),
});

export const proposalInputSchema = z.object({
  title: z.string().min(1, "Title is required"),
  proposalType: z.enum(VALID_PROPOSAL_TYPES).optional().default("general"),
  summary: z.string().optional(),
  content: z.any().optional(),
  agentName: z.string().optional().default("user"),
});

export const campaignMetricsSchema = z.object({
  campaignId: z.string().min(1),
  campaignName: z.string().min(1),
  platform: z.string().min(1),
  impressions: z.number().int().min(0).default(0),
  clicks: z.number().int().min(0).default(0),
  conversions: z.number().int().min(0).default(0),
  cost: z.number().min(0).default(0),
});

export const chatMessageSchema = z.object({
  message: z.string().min(1).max(2000),
});

export type LeadInput = z.infer<typeof leadInputSchema>;
export type ProposalInput = z.infer<typeof proposalInputSchema>;
export type CampaignMetricsInput = z.infer<typeof campaignMetricsSchema>;
